### PR TITLE
Add option to embed documents without tokenizing using OpenAI interface

### PR DESCRIPTION
**Pull request by sudeepag**
This PR updates `embedding_ctx_length` to be optional. If this value is nil, we will perform a batched call to `embed_with_retry` without tokenizing the text. If it is not nil, we will call `_get_len_safe_embeddings` as per usual. This allows us to use the OpenAI interface for embedding documents without having to rely on the OpenAI model tokenizers.

### CODE REVIEW

File: langchain/embeddings/openai.py [hunk 1]
```diff
@@ -473,9 +473,14 @@ def embed_documents(
         Returns:
             List of embeddings, one for each text.
         """
-        # NOTE: to keep things simple, we assume the list may contain texts longer
-        #       than the maximum context and use length-safe embedding function.
-        return self._get_len_safe_embeddings(texts, engine=self.deployment)
```
Comment (baskaryan) [inline langchain/embeddings/openai.py:478]: get_len_safe_embeddings does batching under the hood, this would undo that

Comment (sudeepag) [reply langchain/embeddings/openai.py:478]: Thanks for taking a look, @baskaryan. `_embedding_func` calls `_get_len_safe_embeddings` under the hood, so wouldn't the functionality remain the same for any text longer than `self.embedding_ctx_length`? https://github.com/hwchase17/langchain/blob/3e90c9798828c083445e7a57eee6d00256b9b569/langchain/embeddings/openai.py#L427-L443

Comment (sudeepag) [reply langchain/embeddings/openai.py:478]: We need to be able to call the OpenAI endpoint by passing in text instead of tokens, and we are not currently able to do so with the `_get_len_safe_embeddings` function. Open to any alternative solutions.

Comment (baskaryan) [reply langchain/embeddings/openai.py:478]: does embed_query or embed_documents not work?

Comment (sudeepag) [reply langchain/embeddings/openai.py:478]: `embed_documents` is what I'm modifying here. `embed_query` will not work, since the vector stores are calling `embed_documents` under the hood: https://github.com/hwchase17/langchain/blob/master/langchain/vectorstores/chroma.py#L171

Comment (baskaryan) [reply langchain/embeddings/openai.py:478]: yea but here it sends tokens (potentially across texts) up in batches

https://github.com/hwchase17/langchain/blob/master/langchain/embeddings/openai.py#L357-L363

out of curiosity, what's the use case?

Comment (sudeepag) [reply langchain/embeddings/openai.py:478]: Right, but isn't that only necessary for longer pieces of text? The functionality for longer text would remain the same. The only change being made here is for shorter texts, which I don't suppose require batching anyway?

The use case is https://github.com/mlc-ai/mlc-llm/pull/551. We are enabling document QA for a wide variety of models that are run locally through an OpenAI compatible REST interface. We can't use the tokenized text, since the tokenizer assumes we're using the OpenAI models. Please let me know if I can provide any additional info!

Comment (baskaryan) [reply langchain/embeddings/openai.py:478]: why wouldn't sending a batched call of many short texts be better than sending them one by one?

use case makes sense! but like you said _embedding_func can still call _get_len_safe_embeddings (unless you're setting embedding_ctx_length to be very high?) so does this fix?

what if we made using _get_len_safe_embeddings optional by making `embedding_ctx_length: Optional[int]` and if it's set to None then we change embed_documents to be something like

```python
        if self.embedding_cx_length:
          return self._get_len_safe_embeddings(texts, engine=self.deployment)
               _chunk_size = chunk_size or self.chunk_size

        if self.show_progress_bar:
            try:
                import tqdm

                _iter = tqdm.tqdm(range(0, len(documents), _chunk_size))
            except ImportError:
                _iter = range(0, len(documents), _chunk_size)
        else:
            _iter = range(0, len(documents), _chunk_size)

        for i in _iter:
            response = embed_with_retry(
                self,
                input=documents[i : i + _chunk_size],
                **self._invocation_params,
            )
            batched_embeddings += [r["embedding"] for r in response["data"]]

         for i in range(len(texts)):
            _result = results[i]
            if len(_result) == 0:
                average = embed_with_retry(
                    self,
                    input="",
                    **self._invocation_params,
                )[
                    "data"
                ][0]["embedding"]
            else:
                average = np.average(_result, axis=0, weights=num_tokens_in_batch[i])
            batched_embeddings[i] = (average / np.linalg.norm(average)).tolist()
```

basically implement a batched call without tokenizing

Comment (sudeepag) [reply langchain/embeddings/openai.py:478]: @baskaryan that sounds good to me, but would it be okay to make `embedding_ctx_length` None by default? The Chroma vector store, which will be calling this function, won't let us pass in any additional arguments:

https://github.com/langchain-ai/langchain/blob/master/libs/langchain/langchain/vectorstores/chroma.py#L175

Comment (sudeepag) [reply langchain/embeddings/openai.py:478]: That sounds good to me. I'll update the PR!

**Comment by sudeepag** [comment]
@baskaryan I've updated the PR with your suggestion, and confirmed that it is working successfully for our use case in https://github.com/mlc-ai/mlc-llm/pull/551. Please take a look when you get the chance!

**Comment by sudeepag** [comment]
Hi @baskaryan, gentle nudge requesting a review on this. Please let me know if I can make any additional changes. Thank you!

File: libs/langchain/tests/integration_tests/embeddings/test_openai.py [hunk 2]
```diff
@@ -71,6 +71,38 @@ def test_openai_embedding_with_empty_string() -> None:
...
+
+
+@pytest.mark.asyncio
+async def test_with_without_tokenizing_equal() -> None:
+    documents = ["foo bar baz bum buzz boom"]
+    with_tokenizing = OpenAIEmbeddings()
+    without_tokenizing = OpenAIEmbeddings(embedding_ctx_length=None)
+    tokenize_output = with_tokenizing.embed_documents(documents)
+    text_output = without_tokenizing.embed_documents(documents)
+    assert np.isclose(tokenize_output, text_output, atol=1e-4).sum() > 0.99
```
Comment (baskaryan) [inline libs/langchain/tests/integration_tests/embeddings/test_openai.py:99]: admit this isn't great, shouldn't be an integration test. but we'll need to do lots of cleanup before we run them in CI anyways, so doesn't matter too much for now imo

File: libs/langchain/tests/integration_tests/embeddings/test_openai.py [hunk 3]
```diff
@@ -71,11 +71,51 @@ def test_openai_embedding_with_empty_string() -> None:
...
+
+
+@pytest.mark.asyncio
+async def test_with_without_tokenizing_equal() -> None:
+    documents = ["foo bar baz bum buzz boom"]
+    with_tokenizing = OpenAIEmbeddings()
+    without_tokenizing = OpenAIEmbeddings(embedding_ctx_length=None)
+    tokenize_output = with_tokenizing.embed_documents(documents)
+    text_output = without_tokenizing.embed_documents(documents)
+    assert np.isclose(tokenize_output, text_output, atol=1e-4).mean() > 0.95
```
Comment (baskaryan) [inline libs/langchain/tests/integration_tests/embeddings/test_openai.py:99]: this actually rarely passes, can't tell if it's bc tokenizing changes behavior or because of a bug (doesn't seem to be just randomness bc other tests don't fail as much)

**Comment by sudeepag** [comment]
Hi @baskaryan, could we please merge this if this is good to go? Thanks.

**Comment by KotlinFactory** [comment]
Looks good to me!

**Comment by sudeepag** [comment]
Hi @baskaryan, is this good to go? Can I make any further changes or do testing? Thanks.

**Comment by tqchen** [comment]
Gentle nudge @baskaryan :) Would love to get it in.

**Comment by Sing-Li** [comment]
@baskaryan this PR will greatly increase the adoption of langchain by our active community, pls merge soon 🙏

File: libs/langchain/tests/integration_tests/embeddings/test_openai.py (hunk 3 above)
Comment (hwchase17) [reply libs/langchain/tests/integration_tests/embeddings/test_openai.py:101]: this is also failing for me

**Comment by sudeepag** [comment]
@hwchase17 @baskaryan The test failure is unrelated to this PR. I confirmed this by checking out the current `master` branch and running this several times:

```python
from langchain.embeddings.openai import OpenAIEmbeddings
documents = ["foo bar baz bum buzz boom"]
with_tokenizing = OpenAIEmbeddings()
tokenize_output = with_tokenizing.embed_documents(documents)
print(f"tokenize_output, sum {np.sum(tokenize_output)}")
```

You will observe that the output varies across different runs:
```console
$ python test.py
tokenize_output, sum -1.162553171971551
$ python test.py
tokenize_output, sum -1.1606533408200788
```

The test is failing whenever the embedding generated with `with_tokenizing` differs from the embedding generated with `without_tokenizing`. Could we please merge this PR and investigate why the embedding is varying across runs in a separate issue? I can remove the faulty test from this PR. Thanks.

File: libs/langchain/tests/integration_tests/embeddings/test_openai.py (hunk 3 above)
Comment (sudeepag) [reply libs/langchain/tests/integration_tests/embeddings/test_openai.py:110]: @hwchase17 @baskaryan please see my comment below https://github.com/langchain-ai/langchain/pull/7815/#issuecomment-1685416343

**Comment by sudeepag** [comment]
@hwchase17 @baskaryan I've removed the failing test, could you please take another look?

**Comment by leo-gan** [comment]
@sudeepag Hi , could you, please, resolve the merging issues? After that ping me and I push this PR for the review. Thanks!

**Comment by simmoncn** [comment]
Does the PR has merged to main branch? @sudeepag or has alternative solution? I also meet the same problems, thanks

**Comment by efriis** [comment]
Hey team! I'll close this and start a github discussion, and then happy to migrate it to `langchain-openai` if it makes sense. Just want to make sure I'm understanding the motivation for skipping tokenization!
